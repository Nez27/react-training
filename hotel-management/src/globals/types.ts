type TUser = {
  id: number;
  name: string;
  identifiedCode: string;
  address: string;
  phone: string;
  roomId: number;
};

type TRoom = {
  id: number;
  name: string;
  amount: number;
  price: number;
  discount: number;
  description: string;
  status: boolean;
};

type TStateSchema = {
  [key: string]: {
    value?: string;
    error?: string;
  };
};

type TKeyValue = {
  [key: string]: boolean | undefined | string;
};

type TKeyString = {
  [key: string]: string;
};

type TPropValues = 'value' | 'error' | boolean;

type TValidator = {
  [key: string]: {
    required?: boolean;
    validator?: {
      func?: (value: string) => boolean;
      error?: string;
    };
  };
};

type TResponse = {
  statusCode: number;
  msg: string;
};

export type {
  TUser,
  TRoom,
  TStateSchema,
  TKeyValue,
  TKeyString,
  TPropValues,
  TValidator,
  TResponse,
};
