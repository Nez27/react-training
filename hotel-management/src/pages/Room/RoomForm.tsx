import styled from 'styled-components';

// Styled
import Button from '@commonStyle/Button.ts';
import Input from '@commonStyle/Input.ts';

// Types
import { IRoom } from '@type/rooms.ts';

// Constants
import {
  INVALID_FIELD,
  REQUIRED_FIELD_ERROR,
} from '@constant/formValidateMessage.ts';
import { REGEX } from '../../constants/commons.ts';

// Helpers
import { isValidRegex, isValidString } from '@helper/validators.ts';

// Components
import Form from '@component/Form/index.tsx';

// Hooks
import { useForm } from 'react-hook-form';
import { useCreateRoom } from '@hook/rooms/useCreateRoom.ts';
import { useUpdateRoom } from '@hook/rooms/useUpdateRoom.ts';

const FormBtn = styled(Button)`
  width: 100%;

  &:disabled,
  &[disabled] {
    background-color: var(--disabled-btn-color);
    cursor: no-drop;
  }
`;

interface IRoomFormProp {
  onCloseModal?: () => void;
  room?: IRoom;
}

const RoomForm = ({ onCloseModal, room }: IRoomFormProp) => {
  const { isCreating, createRoom } = useCreateRoom();
  const { isUpdating, updateRoom } = useUpdateRoom();
  const isLoading = isCreating || isUpdating;
  const { id: editId, ...editValues } = { ...room };
  const formMethods = useForm<IRoom>({
    defaultValues: editId 
      ? editValues 
      : {},
  });
  const {
    register,
    handleSubmit,
    reset,
    formState: { errors, isDirty, isValid },
    trigger,
  } = formMethods;

  // Submit form
  const onSubmit = async (newRoom: IRoom) => {
    if (!editId) {
      // Add request
      createRoom(newRoom, {
        onSuccess: () => {
          reset();
          onCloseModal?.();
        },
      });
    } else {
      // Edit request
      newRoom.id = editId!;
      updateRoom(newRoom, {
        onSuccess: () => {
          reset();
          onCloseModal?.();
        },
      });
    }
  };

  return (
    <Form onSubmit={handleSubmit(onSubmit)}>
      <Form.Row label="Name" error={errors?.name?.message}>
        <Input
          type="text"
          id="name"
          {...register('name', {
            required: REQUIRED_FIELD_ERROR,
            validate: {
              checkValidRoomName: (value) =>
                isValidString(value) || INVALID_FIELD,
            },
            onChange: () => trigger('name'),
          })}
        />
      </Form.Row>

      <Form.Row label="Price" error={errors?.price?.message}>
        <Input
          type="text"
          id="price"
          {...register('price', {
            valueAsNumber: true,
            required: REQUIRED_FIELD_ERROR,
            validate: {
              checkPrice: (v) =>
                isValidRegex(new RegExp(REGEX.NUMBER), v.toString()) ||
                INVALID_FIELD,
            },
            onChange: () => trigger('price'),
          })}
        />
      </Form.Row>

      <Form.Action>
        <FormBtn
          type="submit"
          name="submit"
          disabled={!isDirty || !isValid || isLoading}
        >
          {
            !editId 
              ? 'Add' 
              : 'Save'
          }
        </FormBtn>
        <FormBtn type="button" variations="secondary" onClick={onCloseModal}>
          Close
        </FormBtn>
      </Form.Action>
    </Form>
  );
};

export default RoomForm;
