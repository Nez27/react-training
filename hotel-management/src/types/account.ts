interface IAccount {
  email?: string;
  fullName?: string;
  password?: string;
}

export type { IAccount };
