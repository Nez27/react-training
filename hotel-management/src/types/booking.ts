interface IBooking {
  id: number;
  startDate: string;
  endDate: string;
  status: boolean;
  amount: number;
  roomId: number;
  userId: number;
}

interface TBookingResponse {
  id: number;
  startDate: string;
  endDate: string;
  status: boolean;
  amount: number;
  rooms: {
    id: number
    name: string;
  } | null;
  users: {
    id: number;
    name: string;
  } | null;
}

export type { IBooking, TBookingResponse };
