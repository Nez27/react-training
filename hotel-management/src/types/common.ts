type Nullable<T> = T | null;

interface IDataState {
  id: number;
  name?: string;
  status?: boolean;
  isBooked?: boolean;
}

export type { Nullable, IDataState };
