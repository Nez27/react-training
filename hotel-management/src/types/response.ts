type TResponse<T> = {
  statusCode: number;
  msg: string;
  data?: T;
};

export type { TResponse };
