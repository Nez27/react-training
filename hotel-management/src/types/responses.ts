interface IResponse<T> {
  statusCode: number;
  msg: string;
  data?: T;
}

export type { IResponse };
