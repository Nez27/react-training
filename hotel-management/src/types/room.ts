interface IRoom {
  id: number;
  name: string;
  price: number;
  status: boolean;
  isDelete: boolean;
}

export type { IRoom };
