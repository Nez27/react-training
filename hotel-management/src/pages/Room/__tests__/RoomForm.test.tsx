import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { BrowserRouter } from 'react-router-dom';
import {
  RenderResult,
  fireEvent,
  render,
  waitFor,
} from '@testing-library/react';

// Types
import { IRoom } from '@src/types/room';

// Components
import RoomForm from '../RoomForm';

const mockCreateRoom = jest.fn();
const mockUpdateRoom = jest.fn();

jest.mock('@src/hooks/rooms/useCreateRoom.ts', () => ({
  ...jest.requireActual('@src/hooks/rooms/useCreateRoom.ts'),
  useCreateRoom: jest.fn(() => ({
    isCreating: false,
    createRoom: mockCreateRoom,
  })),
}));

jest.mock('@src/hooks/rooms/useUpdateRoom', () => ({
  ...jest.requireActual('@src/hooks/rooms/useUpdateRoom'),
  useUpdateRoom: jest.fn(() => ({
    isUpdating: false,
    updateRoom: mockUpdateRoom,
  })),
}));

let result: {
  name: string;
  price: number;
};


describe('RoomForm', () => {
  const queryClient = new QueryClient();

  let wrapper: RenderResult | null = null;

  test('Should render correctly', () => {
    wrapper = render(
      <QueryClientProvider client={queryClient}>
        <BrowserRouter>
          <RoomForm />
        </BrowserRouter>
      </QueryClientProvider>
    );

    expect(wrapper).toMatchSnapshot();
  });

  test('Should add correctly', async () => {
    wrapper = render(
      <QueryClientProvider client={queryClient}>
        <BrowserRouter>
          <RoomForm />
        </BrowserRouter>
      </QueryClientProvider>
    );

    const nameField = wrapper!.container.querySelector('#name');
    const priceField = wrapper!.container.querySelector('#price');
    const submitBtn = wrapper!.getByText('Add').closest('button');

    mockCreateRoom.mockImplementationOnce((newRoom, { onSuccess }) => {
      result = newRoom;
      onSuccess();
    });

    fireEvent.input(nameField!, { target: { value: 'Room 1' } });
    fireEvent.input(priceField!, { target: { value: '120' } });

    fireEvent.submit(submitBtn!);

    await waitFor(() => {
      expect(mockCreateRoom).toHaveBeenCalled();
      expect(result).toEqual({ name: 'Room 1', price: 120 });
    });
  });

  test('Should edit correctly', async () => {
    const tempRoom: IRoom = {
      id: 1,
      name: 'Room 1',
      price: 2300,
      status: true,
      isDelete: true,
    }

    wrapper = render(
      <QueryClientProvider client={queryClient}>
        <BrowserRouter>
          <RoomForm room={tempRoom}/>
        </BrowserRouter>
      </QueryClientProvider>
    );

    const submitBtn = wrapper!.getByText('Save').closest('button');

    mockUpdateRoom.mockImplementationOnce((newRoom, { onSuccess }) => {
      result = newRoom;
      onSuccess();
    });

    fireEvent.submit(submitBtn!);

    await waitFor(() => {
      expect(mockUpdateRoom).toHaveBeenCalled();
      expect(result).toEqual(tempRoom);
    });
  });
});
