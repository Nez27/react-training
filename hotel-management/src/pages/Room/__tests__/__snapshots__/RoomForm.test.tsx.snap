// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomForm Should render correctly 1`] = `
<form
  className="sc-dlniIP bCPNqw"
  onSubmit={[Function]}
>
  <div
    className="sc-eCAqax eozPUK"
  >
    <label
      className="sc-jSFipO cMYxin"
    >
      Name
    </label>
    <div>
      <input
        className="sc-gtsqUy fGZPSU"
        id="name"
        name="name"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
      />
    </div>
  </div>
  <div
    className="sc-eCAqax eozPUK"
  >
    <label
      className="sc-jSFipO cMYxin"
    >
      Price
    </label>
    <div>
      <input
        className="sc-gtsqUy fGZPSU"
        id="price"
        name="price"
        onBlur={[Function]}
        onChange={[Function]}
        type="text"
      />
    </div>
  </div>
  <div
    className="sc-hKFymg dteuGK"
  >
    <button
      className="sc-bdnyFh sc-iCoHzw hHmoPp jAHgeA"
      disabled={false}
      name="submit"
      type="submit"
      variations="primary"
    >
      Add
    </button>
    <button
      className="sc-bdnyFh sc-iCoHzw bjdOKm jAHgeA"
      type="button"
      variations="secondary"
    >
      Close
    </button>
  </div>
</form>
`;
